xscen/aggregate.py

```diff
@@ -467,3 +491,282 @@ def spatial_mean(
+        )
+        return None
+
+    logger.info(
+        f"Computing warming level +{wl}C for id: {id_ds} from column: {right_column.columns[0]}."
+    )
+
+    # compute reference temperature for the warming
+    mean_base = right_column.loc[tas_base_period[0] : tas_base_period[1]].mean()
+
+    yearly_diff = right_column - mean_base  # difference from reference
+    last_year = right_column.iloc[-1].name  # last available year
+
+    # get the star and end date of the window when the warming level is first reached
+    for year in np.arange(1850, 2101):
+        start_yr = int(year - (window / 2 - 1))
+        end_yr = int(year + window / 2)
+        if end_yr > last_year:  # put nan if end_yr is not available in the data
+            start_yr = np.nan
+            end_yr = np.nan
+            break
+        rolling_mean = yearly_diff.loc[start_yr:end_yr].mean().values
+        if rolling_mean >= wl:
+            break
+        else:  # if it never goes above wl
+            start_yr = np.nan
+            end_yr = np.nan
+
+    if np.isnan(start_yr):
+        logger.info(
+            f"Global warming level of +{str(wl)}C is never reached for {id_ds}."
+        )
+        return None
+
+    # cut the window selected above
+    ds_wl = ds.sel(time=slice(str(start_yr), str(end_yr))).load()
+
+    # compute indicators
+    ind_dict = compute_indicators(ds=ds_wl, indicators=indicators)
+
+    # Compute the window-year mean
+    ds_merge = xr.Dataset()
+    concats = []
+    for freq, ds_ind in ind_dict.items():
+        ds_mean = climatological_mean(
+            ds_ind.dropna(
+                dim="time", how="all"
+            ),  # to drop first and last nan DJF, if not we start a year too early
+            window=window,
+            min_periods=min_window,
+            to_level=to_level.format(wl=wl),
+        )
+
+        if "AS" not in freq:  # there is only one time, can't infer_freq
+            # name new_dim
+            if "QS" in freq:
+                new_dim = "season"
+            elif "MS" in freq:
+                new_dim = "month"
+            else:
+                new_dim = "period"
+                warnings.warn(
+                    f"Frequency {freq} is not supported. Setting name of the new dimension to 'period'."
+                )
+            ds_mean = unstack_dates(
+                ds_mean,
+                # seasons = seasons,
+                new_dim=new_dim,
+            )
+            horizon = ds_mean.horizon.values[0, 0]
+            ds_mean = (
+                ds_mean.drop_vars("horizon")
+                .assign_coords(horizon=("time", [horizon]))
+                .swap_dims({"time": "horizon"})
+                .drop_vars("time")
+            )
+
+        else:
+            ds_mean = ds_mean.swap_dims({"time": "horizon"}).drop_vars("time")
+
+        ds_mean["horizon"] = [f"+{str(wl)}C"]
+        concats.append(ds_mean)
+
+        # put all indicators in one dataset
+        for var in ds_mean.data_vars:
```

Comment:
This could be moved to `climatological_mean`.

Reply:
like passing the dict from `compute_indicators` as input to `climatological_mean` ?

Reply:
(With previous comment too) More so the idea of taking all indicators no matter the frequency, unstacking the time, then putting everything in a single dataset.

Actually, this could maybe be its own function in `indicators.py`?

Reply:
as discussed this stays in produce_horizon only